cpp/dolfinx/mesh/Topology.cpp

```diff
@@ -19,6 +19,105 @@
 using namespace dolfinx;
 using namespace dolfinx::mesh;
 
+namespace
+{
+//-----------------------------------------------------------------------------
+std::unordered_map<std::int64_t, std::vector<int>>
+compute_sharing(MPI_Comm comm, std::vector<std::int64_t>& unknown_indices)
+{
+  const int mpi_size = dolfinx::MPI::size(comm);
+  const std::int64_t global_space
+      = dolfinx::MPI::max(comm, *std::max_element(unknown_indices.begin(),
+                                                  unknown_indices.end()))
+        + 1;
+
+  std::vector<std::vector<std::int64_t>> send_indices(mpi_size);
+  for (std::int64_t global_i : unknown_indices)
+  {
+    const int index_owner
+        = dolfinx::MPI::index_owner(mpi_size, global_i, global_space);
+    send_indices[index_owner].push_back(global_i);
+  }
+
+  std::vector<std::vector<std::int64_t>> recv_indices(mpi_size);
+  dolfinx::MPI::all_to_all(comm, send_indices, recv_indices);
+
+  // Get index sharing - first vector entry (lowest) is owner.
+  std::unordered_map<std::int64_t, std::vector<int>> index_to_owner;
+  for (int p = 0; p < mpi_size; ++p)
+  {
+    const std::vector<std::int64_t>& recv_p = recv_indices[p];
+    for (std::size_t j = 0; j < recv_p.size(); ++j)
+      index_to_owner[recv_p[j]].push_back(p);
+  }
+
+  // Send index ownership data back to all sharing processes
+  std::vector<std::vector<int>> send_owner(mpi_size);
+  for (int p = 0; p < mpi_size; ++p)
+  {
+    const std::vector<std::int64_t>& recv_p = recv_indices[p];
+    for (std::size_t j = 0; j < recv_p.size(); ++j)
+    {
+      const auto it = index_to_owner.find(recv_p[j]);
+      assert(it != index_to_owner.end());
+      const std::vector<int>& sharing_procs = it->second;
+      send_owner[p].push_back(sharing_procs.size());
+      for (int sp : sharing_procs)
+        send_owner[p].push_back(sp);
+    }
+  }
+
+  std::vector<std::vector<int>> recv_owner(mpi_size);
+  dolfinx::MPI::all_to_all(comm, send_owner, recv_owner);
```

Comment:
Any way to avoid this, even if just a comment at this stage pointing to a possible solution.